retriever/lib/scripts.py

```diff
@@ -107,7 +108,101 @@ def get_script(dataset):
+            if r.status_code == 404:
+                return None
+        with open(newpath, 'wb') as f:
+            for chunk in r.iter_content(chunk_size=1024):
+                f.write(chunk)
+        r.close()
+        if is_json:
+            read_script = read_json(join(SCRIPT_WRITE_PATH, script))
+            setattr(read_script, "_file", os.path.join(SCRIPT_WRITE_PATH, script_name))
+            setattr(read_script, "_name", script)
+            return read_script
+        else:
+            file, pathname, desc = imp.find_module(script, [SCRIPT_WRITE_PATH])
+            new_module = imp.load_module(script, file, pathname, desc)
+            setattr(new_module.SCRIPT, "_file", os.path.join(SCRIPT_WRITE_PATH, script_name))
+            setattr(new_module.SCRIPT, "_name", script)
+            return new_module.SCRIPT
+    except:
+        raise
+
+
+def get_dataset_names_upstream(keywords=None, licenses=None, repo=REPOSITORY):
+    """Search all datasets upstream by keywords and licenses."""
+    if not keywords and not licenses:
+        version_file = requests.get(repo + "version.txt").text
+        version_file = version_file.splitlines()[1:]
+
+        scripts = []
+        max_scripts = 100
+        for line in version_file:
+            script = line.strip('\n').split(',')[0]
+            script = '.'.join(script.split('.')[:-1])
+            script = script.replace('_', '-')
+            scripts.append(script)
+            if len(scripts) == max_scripts:
+                break
+        return sorted(scripts)
+
+    result_scripts = set()
+    if repo == RETRIEVER_REPOSITORY:
+        search_url = "https://api.github.com/search/code?q={query}+in:file+path:scripts+repo:weecology/retriever"
+    else:
+        search_url = "https://api.github.com/search/code?q={query}+in:file+path:scripts+repo:weecology/retriever-recipes"
+    if licenses:
+        licenses = [l.lower() for l in licenses]
+        for l in licenses:
+            try:
+                r = requests.get(search_url.format(query=l))
+                r = r.json()
+                for index in range(r['total_count']):
+                    script = r['items'][index]['name']
+                    script = '.'.join(script.split('.')[:-1])
+                    script = script.replace('_', '-')
+                    result_scripts.add(script)
+            except:
+                raise
+    if keywords:
+        keywords = [k.lower() for k in keywords]
+        for k in keywords:
+            try:
+                r = requests.get(search_url.format(query=k))
+                r = r.json()
+                for index in range(r['total_count']):
+                    script = r['items'][index]['name']
+                    script = '.'.join(script.split('.')[:-1])
+                    script = script.replace('_', '-')
+                    result_scripts.add(script)
+            except:
+                raise
+    return sorted(result_scripts)
```

Comment:
If no keywords or licenses argument is provided, it reads the `version.txt` of the according repo and returns an array of script names with a maximum length of 100 to prevent flooding of terminal when doing `retriever ls`. When a query is provided, Github's search api is used (based on initial discussion with @henrykironde) to search scripts with the specified license or citation query parameter. Github's search api can be used to used to search all the script files in the `scripts` folder for the query param. However, it only searches the master branch, and thus may not produce desired results as of now, but would work when the branch is merged into master.